pkg/frontend/openshiftcluster_listorgetkubernetesobjects_post.go

```diff
@@ -0,0 +1,92 @@
+	"github.com/Azure/ARO-RP/pkg/database/cosmosdb"
+	"github.com/Azure/ARO-RP/pkg/frontend/middleware"
+	"github.com/Azure/ARO-RP/pkg/util/dynclient"
+	"github.com/Azure/ARO-RP/pkg/util/restconfig"
+)
+
+func (f *frontend) postListOrGetKubernetesObjects(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	log.Infof("postListOrGetKubernetesObjects")
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	if len(body) > 0 && !json.Valid(body) {
+		api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content was invalid and could not be deserialized.")
+		return
+	}
+
+	r.URL.Path = filepath.Dir(r.URL.Path)
+
+	b, err := f._postListOrGetKubernetesObjects(ctx, r)
+
+	reply(log, w, nil, b, err)
+}
+
+func (f *frontend) _postListOrGetKubernetesObjects(ctx context.Context, r *http.Request) ([]byte, error) {
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	vars := mux.Vars(r)
+
+	_, err := f.validateSubscriptionState(ctx, r.URL.Path, api.SubscriptionStateRegistered)
+	if err != nil {
+		return nil, err
+	}
+
+	doc, err := f.db.OpenShiftClusters.Get(ctx, r.URL.Path)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return nil, api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
+	case err != nil:
+		return nil, err
+	}
+
+	if doc.OpenShiftCluster.Properties.ProvisioningState == api.ProvisioningStateCreating ||
+		doc.OpenShiftCluster.Properties.ProvisioningState == api.ProvisioningStateDeleting ||
+		doc.OpenShiftCluster.Properties.ProvisioningState == api.ProvisioningStateFailed && doc.OpenShiftCluster.Properties.FailedProvisioningState == api.ProvisioningStateCreating ||
+		doc.OpenShiftCluster.Properties.ProvisioningState == api.ProvisioningStateFailed && doc.OpenShiftCluster.Properties.FailedProvisioningState == api.ProvisioningStateDeleting {
+		return nil, api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeRequestNotAllowed, "", "Request is not allowed in provisioningState '%s'.", doc.OpenShiftCluster.Properties.ProvisioningState)
+	}
+
+	var parameters v20191231preview.OpenShiftClusterListOrGetKubernetesObjectsParameters
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	err = json.Unmarshal(body, &parameters)
+	if err != nil {
+		return nil, api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content was invalid and could not be deserialized: %q.", err)
+	}
+
+	config, err := restconfig.RestConfig(f.env, doc.OpenShiftCluster)
+	if err != nil {
+		return nil, err
+	}
+
+	dc, err := dynclient.New(log, config)
+	if err != nil {
+		return nil, err
+	}
+	ul, err := dc.GetOrList(ctx, parameters.Kind, parameters.Namespace, parameters.Name)
```

Comment:
Im a little bit worried on overall this implementation. This should not be in the public customer api. Reason being that if the customer has access to Azure API, but OpenShift administrator granted him/her only view role to the OCP cluster. This basically opens a path for users to access the cluster objects other way.

Reply:
I think we need a better way to split methods for different apis. Like we will be adding new Methods to new API versions, and not backporting. This is very similar usecase.

Reply:
I agree with the above. I think, the plan is to have to have the `/admin` prefix for admin API. So one option to separate things would be to have two routers (see `pkg/frontend/frontend.go`):

* A router for costumer API similar to what we already have + in a middleware forbid access to the `admin` API version via this router
* A router for admin API. It will include things from the customer API + in a middleware forbid access everything that is != `admin` API version

Reply:
FYI, we already have a piece of code that deals with access to admin/non-admin api versions.

https://github.com/Azure/ARO-RP/blob/f205c8a532555fe5899fe910973f625cded15563/pkg/frontend/middleware/authenticated.go#L22-L33